libplatsupport/src/plat/polarfire/ltimer.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * Copyright 2019, Data61
+ * Commonwealth Scientific and Industrial Research Organisation (CSIRO)
+ * ABN 41 687 119 230.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+/*
+ * Implementation of a logical timer for Microchip Polarfire Icicle platform.
+ */
+#include <platsupport/timer.h>
+#include <platsupport/ltimer.h>
+#include <platsupport/pmem.h>
+#include <platsupport/plat/mstimer.h>
+#include <utils/util.h>
+
+#include "../../ltimer.h"
+
+enum {
+    TIMEOUT_TIMER,
+    TIMESTAMP_TIMER,
+    NUM_TIMERS
+};
+
+#define TIMEOUT_ADDR_OFFSET   MSTIMER1_OFFSET
+#define TIMESTAMP_ADDR_OFFSET MSTIMER2_OFFSET
+
+
+typedef struct {
+    void *base_address;
+    mstimer_t timeout;
+    mstimer_t timestamp;
+    ps_io_ops_t ops;
+    timer_callback_data_t callback_data[NUM_TIMERS];
+    ltimer_callback_fn_t user_callback;
+    void *user_callback_token;
+} polarfire_timers_t;
+
+static ps_irq_t irqs[] = {
+    {
+        /* Timeout */
+        .type = PS_INTERRUPT,
+        .irq.number = MSTIMER_TIMER1INT
+    },
+    {
+        /* Timestamp */
+        .type = PS_INTERRUPT,
+        .irq.number = MSTIMER_TIMER2INT
+    },
+};
+
+static pmem_region_t timer_pmem = {
+    .type = PMEM_TYPE_DEVICE,
+    .base_addr = MSTIMER1_PADDR,
+    .length = PAGE_SIZE_4K
+};
+
+
+#define N_IRQS ARRAY_SIZE(irqs)
+#define N_PMEMS 1
+
+size_t get_num_irqs(void *data)
+{
+    return N_IRQS;
+}
+
+static int get_nth_irq(void *data, size_t n, ps_irq_t *irq)
+{
+    if (n >= N_IRQS) {
+        ZF_LOGE("IRQ requested out of range");
+        return EINVAL;
+    }
+
+    *irq = irqs[n];
+    return 0;
+}
+
+static size_t get_num_pmems(void *data)
+{
+    return N_PMEMS;
+}
+
+static int get_nth_pmem(void *data, size_t n, pmem_region_t *paddr)
+{
+    if (n == 0) {
+        *paddr = timer_pmem;
+    } else {
+        ZF_LOGE("PMEM requested out of range");
+        return EINVAL;
+    }
+    return 0;
+}
+
+static int  ltimer_handle_irq(void *data, ps_irq_t *irq)
+{
+    /* Ack irq */
+    assert(data != NULL);
+
+    polarfire_timers_t *timers = data;
+    long irq_number = irq->irq.number;
+    ltimer_event_t event;
+
+    if (irq_number == irqs[TIMESTAMP_TIMER].irq.number) {
+        mstimer_handle_irq(&timers->timestamp, irq_number);
+        event = LTIMER_OVERFLOW_EVENT;
+    } else if (irq_number == irqs[TIMEOUT_TIMER].irq.number) {
+        mstimer_handle_irq(&timers->timeout, irq_number);
+        event = LTIMER_TIMEOUT_EVENT;
+    } else {
+        ZF_LOGE("Invalid IRQ number: %d received.\n", irq_number);
```

Comment:
```suggestion
        ZF_LOGE("Invalid IRQ number: %ld received.\n", irq_number);
```
Since `irq_number` is a long, we should use `%ld` here.

Reply:
Done